src/cli/BeamExtract.py

```diff
@@ -0,0 +1,287 @@
+            out_p = out_it.particles[k_p]
+
+            num_particles = in_p["momentum"]["x"].shape[0]
+            print(f"   Number of particles: {num_particles}")
+
+            N_pass = int(np.ceil(num_particles/slice_size))
+            print(f"   number of passes: {N_pass}")
+
+            # bookkeeping in the global output arrays of the species
+            out_slice_start = 0
+            out_slice_end = 0
+
+            # stepping through particle in slices
+            for slice_start in range(0, num_particles, slice_size):
+                slice_end = slice_start + slice_size
+                if slice_end > num_particles:
+                    slice_end = num_particles
+                #print('    {:,} {:,}'.format(slice_start, slice_end))
+                in_slice = np.s_[slice_start:slice_end]
+
+                # species record: filter
+                accepted = filter_species(in_p, in_slice)
+                out_slice_end = out_slice_start + np.sum(accepted)
+
+                if out_slice_end == out_slice_start:
+                    continue
+
+                # species records
+                for k_p_r, in_p_r in in_p.items():
+                    print(f"    {k_p_r}")
+
+                    # species record components: data
+                    for k_p_rc, in_p_rc in in_p_r.items():
+                        print(f"      {k_p_rc} {in_p_rc.shape}->[{slice_start}:{slice_end}]")
+
+                        # copy data
+                        if in_p_rc.empty:
+                            out_p_r = out_p[k_p_r]
+                            out_p_rc = out_p_r[k_p_rc]
+                            if not out_p_rc.empty:
+                                dataset = io.Dataset(in_p_rc.dtype, (0, ))
+                                dataset.options = json.dumps(dataset_config)
+                                out_p_rc.reset_dataset(dataset)
+                                # out_p_rc.make_empty(dtype, 1)  # done by reset_datatype w/ zero shape already
+                        elif in_p_rc.constant:
+                            # later, once we know the final shape
+                            pass
+                        else:
+                            data = in_p_rc[in_slice]
+                            input_series.flush()
+
+                            # write accepted particles back
+                            out_slice = np.s_[out_slice_start:out_slice_end]
+                            print(f"        out_slice={out_slice}, {out_slice_start}")
+
+                            out_p_r = out_p[k_p_r]
+                            out_p_rc = out_p_r[k_p_rc]
+                            dataset = io.Dataset(in_p_rc.dtype, (out_slice_end,))
+                            dataset.options = json.dumps(dataset_config_resizable)
+                            out_p_rc.reset_dataset(dataset)
+                            out_p_rc[out_slice] = data[accepted]
+                            output_series.flush()
```

Comment:
We could avoid OOM for the output of the while step by using:
- BP5 and adding
- flush_target="disk"

Reply:
Looking into the documentation on this, I notice a documentation bug.
https://openpmd-api.readthedocs.io/en/0.15.1/backends/adios2.html mentions `adios2.preferred_flush_target = "disk"` while https://openpmd-api.readthedocs.io/en/0.15.1/details/backendconfig.html talks of `adios2.engine.preferred_flush_target = "disk"`.
The correct one is **with** engine, so: `series.flush("""adios2.engine.preferred_flush_target = "disk" """)`